### ISSUE TITLE
Support pluralization

### ISSUE THREAD

**Issue by sokra** [webpack-contrib/i18n-webpack-plugin#1]
https://developer.mozilla.org/en-US/docs/Localization_and_Plurals

``` javascript
__(count, ["one item", "#1 items"], "nItems").replace("#1", count)
```

``` javascript
{ // German
  nItems: ["Ein Element", "#1 Elemente"]
}
```

**Comment by max-mykhailenko** [comment]
I will try to combine current loader with external solution http://i18next.com/pages/doc_features.html

**Comment by ronkorving** [comment]
That might be a great solution.

**Comment by max-mykhailenko** [comment]
@ronkorving It will be available on this weekends. Only small improvements left.

**Comment by ronkorving** [comment]
Are you saying you almost have a radically new version of i18-webpack-plugin ready? That depends on [i18next-node](https://github.com/i18next/i18next-node)?

**Comment by max-mykhailenko** [comment]
@ronkorving Yes. Webpack generate bundle for each language using i18next-node. Static text already replaced completely as in current plugin but with additional i18next options. Plurals we must left as expression and replace dynamically on client side, Now I working on including language files and i18next module to the bundles. Unfortunately webpack doesn't has documentation for that.

**Comment by ronkorving** [comment]
Yeah, Webpack documentation on authoring plugins is very limited. I really appreciate your work on this, and can't wait to see the result! Thanks a lot, really.

**Comment by martyphee** [comment]
Is this being worked on anymore?

**Comment by max-mykhailenko** [comment]
@martyphee simple language bundles generation done, plurals must work only dynamic with i18next module. At this point I stop development of this feature. I can create gist with example

**Comment by martyphee** [comment]
Sure that would be great.

**Comment by karlhorky** [comment]
@max-mykhailenko I'd also like to see a gist of this

**Comment by max-mykhailenko** [comment]
Sorry for delay https://gist.github.com/max-mykhailenko/64ce006f9febe4f0dbe6

**Comment by martyphee** [comment]
Thank you for the Gist.  It's very helpful

**Comment by karlhorky** [comment]
Thanks for the gist. Are you planning on opening a pull request to add this to the plugin? Or do you have your own version of the plugin that you'll release?

**Comment by max-mykhailenko** [comment]
@karlhorky Actually I don't understand how I can combine `json` generating and webpack building process. Maybe it's loader for files `.php .twig .js' and than using `I18nPlugin`. I will appreciate if you help me with architecture if you have experience with that.

**Comment by karlhorky** [comment]
I'm not sure you need to generate the json yourself - that can be part of a different plugin, loader, etc. For instance, take a look at [the i18n example](https://github.com/webpack/webpack/tree/master/examples/i18n). This plugin just reads in an existing json file.

---

I took a quick look at the plugin code, and it seems to me the correct way to go about it would be this:

1. Define a new function (maybe called `ngettext`?) [near the top](https://github.com/webpack/i18n-webpack-plugin/blob/3d85b5df3778e05c80305660cead44f5cadadf3d/index.js#L30) that will accept the arguments for plural translations. Maybe take inspiration from libraries like [Jed](http://slexaxton.github.io/Jed/)
2. Define a new `parser.plugin` block to call the function and define behavior like [the existing one](https://github.com/webpack/i18n-webpack-plugin/blob/3d85b5df3778e05c80305660cead44f5cadadf3d/index.js#L43-L81).
3. Refactor: see if the existing `parser.plugin` function shares enough with the new function defined in the previous step that some of the code can be shared. Move any shared things out into their own standalone functions.

The question would be if `i18next-node` can help with step 2 or 3 above, or if the code would be simple enough that you just write your own implementation.

---
Or, @sokra @jhnns? Would this be a good way of doing it?

**Comment by karlhorky** [comment]
Alternately, the `__()` function could be reused (polymorphic based on the number and type of arguments passed to it). This is already being done [here](https://github.com/webpack/i18n-webpack-plugin/blob/3d85b5df3778e05c80305660cead44f5cadadf3d/index.js#L45-L61). Not sure if this would make the function do too much though.

**Comment by jhnns** [comment]
@karlhorky I think, the i18n plugin should support Jed style functions. Personally, I'd expect it to handle *proper pluralization* and *contexts*. Contexts are an important feature, because translations might be totally different based on the context they occur. That's probably something that native speaker tend to underestimate. 

So, basically we should provide these functions:

```javascript
gettext = function ( key )
dgettext = function ( domain, key )
ngettext = function ( singular_key, plural_key, value )
dngettext = function ( domain, singular_ley, plural_key, value )
pgettext = function ( context, key )
dpgettext = function ( domain, context, key )
npgettext = function ( context, singular_key, plural_key, value )
dnpgettext = function ( domain, context, singular_key, plural_key, value )
```

I know, the function names are kind of weird, but they are already well-known across different languages.

I think that should be doable with the current implementation. To simplify the implementation, that should probably be a breaking change.

**Comment by karlhorky** [comment]
> that should be doable with the current implementation

So as to [my proposal above](https://github.com/webpack/i18n-webpack-plugin/issues/1#issuecomment-185642294), we would:

1. Define multiple new functions (along with keeping `__()` for backwards compatibility?)
2. Define `parser.plugin` blocks for each of them?
3. Refactor: In your opinion, could any of them share any code? Or is that what you meant by "breaking change"?

Would you leverage a library such as `i18next-node` or `Jed` for use in any of the `plugin.parser` functions or would you suggest we reimplement these functions ourselves?

**Comment by martyphee** [comment]
I would suggest `i18next` since it can handle more formats.  Jed does work
well.  We currently use that.

**Comment by johannesnagl** [comment]
What's the status of this issue? Any progress since last year?
